File: Easy/2. Validate Subsequence/validateSubsequence.py
# O(n) time | O(1) space
def solution1(array, sequence):
    index = 0
    for i in array:
        if index == len(sequence):
            break
        if sequence[index] == i:
            index += 1
    return index == len(sequence)

# O(n) time | O(1) space
def solution2(array, sequence):
    array_index = 0
    sequence_index = 0
    while array_index < len(array) and sequence_index < len(sequence):
        if array[array_index] == sequence[sequence_index]:
            sequence_index += 1
        array_index += 1
    return sequence_index == len(sequence)

if __name__ == "__main__":
    array = [5, 1, 22, 25, 6, -1, 8, 10]
    sequence = [1, 6, -1, 10]

    print(f"Solution 1: {solution1(array, sequence)}")
    print(f"Solution 2: {solution2(array, sequence)}")

File: Easy/2. Validate Subsequence/validateSubsequence.go
package main

import "fmt"

// O(n) time | O(1) space
func solution1(array []int, sequence []int) bool {
	index := 0
	for _, value := range array {
		if index == len(sequence) {
			break
		}
		if sequence[index] == value {
			index += 1
		}
	}
	return index == len(sequence)
}

// O(n) time | O(1) space
func solution2(array []int, sequence []int) bool {
	arrIndex := 0
	seqIndex := 0
	for arrIndex < len(array) && seqIndex < len(sequence) {
		if array[arrIndex] == sequence[seqIndex] {
			seqIndex += 1
		}
		arrIndex += 1
	}
	return seqIndex == len(sequence)
}

func main() {
	array := []int{5, 1, 22, 25, 6, -1, 8, 10}
	sequence := []int{1, 6, -1, 10}

	fmt.Println("Solution 1: ", solution1(array, sequence))
	fmt.Println("Solution 2: ", solution2(array, sequence))
}
